sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java

```diff
@@ -0,0 +1,166 @@
+
+import com.azure.cosmos.implementation.GoneException;
+import com.azure.cosmos.implementation.RxDocumentServiceRequest;
+import com.azure.cosmos.implementation.directconnectivity.IAddressResolver;
+import com.azure.cosmos.implementation.routing.PartitionKeyRangeIdentity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.channels.ClosedChannelException;
+import java.time.Instant;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class RntbdConnectionStateListener {
+    // region Fields
+
+    private static final Logger logger = LoggerFactory.getLogger(RntbdConnectionStateListener.class);
+
+    private final IAddressResolver addressResolver;
+    private final RntbdEndpoint endpoint;
+    private final Set<PartitionKeyRangeIdentity> partitionAddressCache;
+    private final AtomicBoolean updatingAddressCache = new AtomicBoolean(false);
+
+    // endregion
+
+    // region Constructors
+
+    public RntbdConnectionStateListener(final IAddressResolver addressResolver, final RntbdEndpoint endpoint) {
+        this.addressResolver = checkNotNull(addressResolver, "expected non-null addressResolver");
+        this.endpoint = checkNotNull(endpoint, "expected non-null endpoint");
+        this.partitionAddressCache = ConcurrentHashMap.newKeySet();
+    }
+
+    // endregion
+
+    // region Methods
+
+    public void onException(final RxDocumentServiceRequest request, Throwable exception) {
+        checkNotNull(request, "expect non-null request");
+        checkNotNull(exception, "expect non-null exception");
+
+        if (exception instanceof GoneException) {
+            final Throwable cause = exception.getCause();
+
+            if (cause != null) {
+
+                // GoneException was produced by the client, not the server
+                //
+                // This could occur for example:
+                //
+                // * an operation fails due to an IOException which indicates a connection reset by the server,
+                // * a channel closes unexpectedly because the server stopped taking requests, or
+                // * an error was detected by the transport client (e.g., IllegalStateException)
+                // * a request timed out in pending acquisition queue
+                // * a request failed fast in admission control layer due to high load
+                // * channel connect timed out
+                //
+                // Currently, only ClosedChannelException will raise onConnectionEvent since it is more sure of a signal the server is going down.
+
+                if (cause instanceof IOException) {
+                    final Class<?> type = cause.getClass();
+
+                    if (type == ClosedChannelException.class) {
```

Comment:
any reason this is preferred over `cause instance of ClosedChannelException` ?

Reply:
changed to instance of ClosedChannelException